enumeratum-core/compat/src/main/scala-2.11/enumeratum/values/IndexedEnumMap.scala

```diff
@@ -0,0 +1,73 @@
+package enumeratum.values
+
+/**
+ * An [[ValueEnum]] that has an optimized array access that does not uses
+ * boxing when accessing the methods [[ValueEnum.withValue()]] and [[ValueEnum.withValueOpt()]]
+ */
+sealed abstract class IndexedEnumMap[@specialized(Int, Short, Long) ValueType <: AnyVal, EntryType <: ValueEnumEntry[ValueType]](values: Seq[EntryType]) {
+
+  private[this] val existingEntriesString = values.map(_.value).mkString(", ")
+  private[this] val (arrayValues, minValue) = {
+    val indices = values.map(v => toIndex(v.value))
+
+    val minValue = indices.min
+    val maxValue = indices.max
+
+    val retVal = Array.fill[Option[EntryType]](maxValue - minValue + 1)(None)
+
+    for (value <- values) {
+      retVal(toIndex(value.value) - minValue) = Some(value)
```

Comment:
This won't work if the range of values is greater than IntMax. You'll run into integer overflow issues and end up trying to access a negative index of the array.